### DISCUSSION TITLE
Incorrect BIP49 P2SH-P2WPKH address

### DISCUSSION THREAD

**Discussion by 3tbraden** [trustwallet/wallet-core#1342] [Q&A] [answered]
**Describe the bug**
`TWBitcoinAddressCreateWithPublicKey` with `TWCoinTypeP2shPrefix(TWCoinTypeBitcoin)` does not product the correct address. With public key `03b54d899b0760164100601571f783a99a8d456975a70303d14a4828882a118de9`, I got `3FNHEJmrBdBDpQn1PfjB1qhsXWJUeygQzP` instead of `3GMJNpE5Xj44YEJbNskyHLVQkd8DV7GDCE`.

**To Reproduce**
```cpp
TEST(TWBitcoinAddress, CreateP2SHWithPublicKey) {
    const auto pubKey = WRAP(TWPublicKey, TWPublicKeyCreateWithData(DATA("03b54d899b0760164100601571f783a99a8d456975a70303d14a4828882a118de9").get(), TWPublicKeyTypeSECP256k1));
    EXPECT_TRUE(pubKey != nullptr);
    const auto addr = WRAP(TWBitcoinAddress, TWBitcoinAddressCreateWithPublicKey(pubKey.get(), TWCoinTypeP2shPrefix(TWCoinTypeBitcoin)));
    EXPECT_TRUE(addr.get() != nullptr);
    EXPECT_EQ(std::string(TWStringUTF8Bytes(TWBitcoinAddressDescription(addr.get()))), "3GMJNpE5Xj44YEJbNskyHLVQkd8DV7GDCE");
}
```

**Expected behavior**
I used the above code to test BIP49 address generation, public key from the mnemonic `left already pipe` and derivation path `m/49'/0'/0'/0/0`, and expected address to be `3GMJNpE5Xj44YEJbNskyHLVQkd8DV7GDCE` as verified from iancoleman.io and electrum wallet.

**Screenshots**
![image](https://user-images.githubusercontent.com/19732240/112263710-b68db000-8caa-11eb-957c-2b6d6b48d021.png)

**Comment by optout21** [comment]
Creating address from public key is a simple hash, it has nothing to do with mnemonic or derivation path.
Sh prefix for bitcoin is 5.
I don't know how did you derived the expected address.
With the given 3-word mnemonic (invalid, BTW), I don't see the same public key in https://iancoleman.io/bip39/.
Please be more specific.

**Reply by 3tbraden** [reply]
it is the public key for that address which I got from the mnemonic and the derivation path on iancoleman.
![image](https://user-images.githubusercontent.com/19732240/112450384-53297e00-8d8f-11eb-9684-96b5ff9c247d.png)

**Reply by optout21** [reply]
If you check iancoleman, for BIP49 it uses P2wpkhInP2sh; see
https://github.com/iancoleman/bip39/blob/master/src/js/index.js#L1382
Address is not derived from pubkey directly, but from script.
I believe these steps can be also done using wallet-core functions (Script, hash, ...)
Note that for BTC, TrustWallet uses BIP84 format ('bc1...')

**Reply by 3tbraden** [reply]
Thank you for the insight, it's very helpful. So I looked up how BIP49 addresses are generated and come up with the code that I am posting as the answer here(I am actually working with golang). Basically I have to prepare the key hash and script data and generate the address from it with a designated prefix.

**Comment by 3tbraden** [comment, marked as answer]
```go
pubKeyHashData := C.TWHashSHA256RIPEMD(C.TWPublicKeyData(pub))
redeemScriptBytes := []byte{}
redeemScriptBytes = append(redeemScriptBytes, 0x00, 0x00)
redeemScriptBytes[1] = 0x14
redeemScriptBytes = append(redeemScriptBytes, TWDataGoBytes(pubKeyHashData)...)
addrBytes := C.TWHashSHA256RIPEMD(TWDataCreateWithGoBytes(redeemScriptBytes))
data := []byte{}
data = append(data, 0x05)
data = append(data, TWDataGoBytes(addrBytes)...)
p2shp2wpkhAddr := C.TWBitcoinAddressCreateWithData(TWDataCreateWithGoBytes(data))
```
